pkg/build/strategies/sti/sti.go

```diff
@@ -237,27 +243,18 @@ func (b *STI) PostExecute(containerID string, location string) error {
 // Determine determines if the current build supports incremental workflow.
 // It checks if the previous image exists in the system and if so, then it
 // verifies that the save-artifacts script is present.
-func (b *STI) Determine(request *api.Request) (err error) {
-	//request.Incremental = false
-
+func (b *STI) Determine(request *api.Request) bool {
 	if request.Clean {
-		return
+		return false
 	}
 
-	// can only do incremental build if runtime image exists
-	previousImageExists, err := b.docker.IsImageInLocalRegistry(request.Tag)
-	if err != nil {
-		return
+	// can only do incremental build if runtime image exists, so always pull image
+	previousImageExists, _ := b.docker.IsImageInLocalRegistry(request.Tag)
+	if image, _ := b.docker.PullImage(request.Tag); image != nil {
+		previousImageExists = true
 	}
 
-	// we're assuming save-artifacts to exist for embedded scripts (if not we'll
-	// warn a user upon container failure and proceed with a clean build)
-	// for external save-artifacts - check its existence
-	saveArtifactsExists := request.ExternalOptionalScripts ||
-		b.fs.Exists(filepath.Join(request.WorkingDir, "upload", "scripts", string(api.SaveArtifacts)))
-
-	request.Incremental = previousImageExists && saveArtifactsExists
-	return nil
+	return previousImageExists && b.installedScripts[api.SaveArtifacts]
```

Comment:
I'm guessing you mean you have an image with `STI_SCRIPTS_URL` to to eg. `image:///myscripts/bin`. Then if you carefully look through the `install.go` you'll see that:
1. I'll invoke download for `defaultURL` see [here](https://github.com/soltysh/source-to-image/commit/10a9d2e91d5c299b8c0cad287aa3e261ab3c3b0d#diff-24a010a7140e82378231d9b71a31792eR96)
2. The return value for `save-artifacts` will be `NewScriptsInsideImageError` see [here](https://github.com/soltysh/source-to-image/commit/48e76c8aa2399abe587a7dc31d2fb48c8791e25b#diff-0fbae86cd401240768b81a29c45e3d62R116)
3. During installing I'll mark this script as not downloaded but installed see [here](https://github.com/soltysh/source-to-image/commit/10a9d2e91d5c299b8c0cad287aa3e261ab3c3b0d#diff-24a010a7140e82378231d9b71a31792eR149)

Thus you'll get exactly what you want :smile: I hope that clarifies this